README.md

```diff
@@ -1371,6 +1371,364 @@ With all of our builtin logic integrated into the codebase, we can implement any
+
+```go
+type PoseidonBuiltinRunner struct {
+ base     memory.Relocatable
+ included bool
+ cache    map[memory.Relocatable]lambdaworks.Felt
+}
+
+func NewPoseidonBuiltinRunner(included bool) *PoseidonBuiltinRunner {
+ return &PoseidonBuiltinRunner{included: included, cache: make(map[memory.Relocatable]lambdaworks.Felt)}
+}
+```
+
+In order to store it as a `BuiltinRunner` we will have to implement the `BuiltinRunner` interface. Aside from `AddValidationRule` &`DeduceMemoryCell`, most builtins share the same bahaviour in their methods, so we can just port them from the builtin runners we implemented before:
+
+```go
+
+const POSEIDON_BUILTIN_NAME = "poseidon"
+
+func (p *PoseidonBuiltinRunner) Base() memory.Relocatable {
+ return p.base
+}
+
+func (p *PoseidonBuiltinRunner) Name() string {
+ return POSEIDON_BUILTIN_NAME
+}
+
+func (p *PoseidonBuiltinRunner) InitializeSegments(segments *memory.MemorySegmentManager) {
+ p.base = segments.AddSegment()
+}
+
+func (p *PoseidonBuiltinRunner) InitialStack() []memory.MaybeRelocatable {
+ if p.included {
+  return []memory.MaybeRelocatable{*memory.NewMaybeRelocatableRelocatable(p.base)}
+ } else {
+  return nil
+ }
+}
+```
+
+As the poseidon builtin doesn't have validation rules, the method will be left empty:
+
+```go
+func (p *PoseidonBuiltinRunner) AddValidationRule(*memory.Memory) {
+}
+```
+
+Now lets dive into the poseidon builtin's behaviour!
+
+The poseidon builtin memory is divided into instancess of 6 cells, 3 input cells and 3 output cells. This means that whenever we want to deduce the value of an output cell, we will look for the input cells, compute the pedersen permutation over them, and write the permutated values to the output cells. As we only deduce the value of one output cell at a time, we will write the value of the output cells to a cache and use them the next time we have to deduce a memory cell so we avoid computing the poseidon hash more than once over the same input values
+
+We define the following constants to represent a poseidon instance:
+
+```go
+const POSEIDON_CELLS_PER_INSTANCE = 6
+const POSEIDON_INPUT_CELLS_PER_INSTANCE = 3
+```
+
+And we can implement `DeduceMemoryCell`:
+
+This method will first check if the cell is an input cell, if its an input cell then there is nothing to deduce and it returns nil. Then it will check if there is a cached value for that cell and return it. If there is no cached value it will define the addresses of the first input and output cells, and fetch the values of the input cells. If any of the input cells is missing, or is not a felt value, it returns an error. Once it has the three input cells, it performs the poseidon permutation and inserts the permutated value into each output cell's address in the cache. It then returns the value stored in the cache for the address that the method received.
```

Comment:
```suggestion
This method will first check if the cell is an input cell, if it's an input cell then there is nothing to deduce and it returns nil. Then it will check if there is a cached value for that cell and return it. If there is no cached value it will define the addresses of the first input and output cells, and fetch the values of the input cells. If any of the input cells is missing, or is not a felt value, it returns an error. Once it has the three input cells, it performs the poseidon permutation and inserts the permutated value into each output cell's address in the cache. It then returns the value stored in the cache for the address that the method received.
```